src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -1,48 +1,77 @@
-import lombok.RequiredArgsConstructor;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Stream;
+import javax.annotation.CheckForNull;
 import lombok.extern.log4j.Log4j2;
 import org.folio.fqm.exception.EntityTypeNotFoundException;
-import org.folio.fqm.exception.InvalidEntityTypeDefinitionException;
 import org.folio.fqm.repository.EntityTypeRepository;
-import org.folio.querytool.domain.dto.ArrayType;
+import org.folio.fqm.utils.flattening.SourceUtils;
 import org.folio.querytool.domain.dto.EntityType;
 import org.folio.querytool.domain.dto.EntityTypeColumn;
 import org.folio.querytool.domain.dto.EntityTypeSource;
-import org.folio.querytool.domain.dto.EntityTypeSourceJoin;
-import org.folio.querytool.domain.dto.Field;
-import org.folio.querytool.domain.dto.NestedObjectProperty;
-import org.folio.querytool.domain.dto.ObjectType;
+import org.folio.querytool.domain.dto.EntityTypeSourceEntityType;
 import org.folio.spring.FolioExecutionContext;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Service;
 
-import java.util.*;
-import java.util.function.Function;
-import java.util.stream.Stream;
-import java.util.stream.StreamSupport;
-
-@Service
-@RequiredArgsConstructor
 @Log4j2
+@Service
 public class EntityTypeFlatteningService {
+
   private final EntityTypeRepository entityTypeRepository;
-  private final ObjectMapper objectMapper;
   private final LocalizationService localizationService;
   private final FolioExecutionContext executionContext;
   private final UserTenantService userTenantService;
 
+  private final Cache<EntityTypeCacheKey, EntityType> entityTypeCache;
+
+  @Autowired
+  public EntityTypeFlatteningService(
+    EntityTypeRepository entityTypeRepository,
+    LocalizationService localizationService,
+    FolioExecutionContext executionContext,
+    UserTenantService userTenantService,
+    @Value("${mod-fqm-manager.entity-type-cache-timeout-seconds:300}") long cacheDurationSeconds
+  ) {
+    this.entityTypeRepository = entityTypeRepository;
+    this.localizationService = localizationService;
+    this.executionContext = executionContext;
+    this.userTenantService = userTenantService;
+
+    this.entityTypeCache =
+      Caffeine.newBuilder().expireAfterWrite(cacheDurationSeconds, java.util.concurrent.TimeUnit.SECONDS).build();
+  }
+
   public EntityType getFlattenedEntityType(UUID entityTypeId, String tenantId) {
-    return getFlattenedEntityType(entityTypeId, null, tenantId);
+    return entityTypeCache.get(
+      new EntityTypeCacheKey(tenantId, entityTypeId, localizationService.getCurrentLocales()),
+      k -> getFlattenedEntityType(k.entityTypeId(), null, k.tenantId())
+    );
   }
```

Comment:
Why are locales here? The flattened entity types are localized as part of flattening, and we don't want to serve someone a different locale